python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py

```diff
@@ -184,6 +184,53 @@ def inference(self,
 # on_fit_start (LightningModule method overwrite)
 def on_fit_start(self):
     self._ortsess_up_to_date = False
+    self.exit_onnx()
+
```

Comment:
how to handle `nn.Module.train()`?

Reply:
good question, but why would a pl user call `nn.Module.train()`? I will go though some example doc and get back to this comment.

Reply:
> good question, but why would a pl user call `nn.Module.train()`? I will go though some example doc and get back to this comment.

It's common practice to call `model.eval` before inference, and call `model.train` before training, no?

Reply:
For inference yes, since `Trainer.predict` (internally call `.eval()`) is not used frequently. Fore training no, `model.train()` is not required and not used in pl docs and users only needs to call `Trainer.fit`.

Reply:
Not for PL; but it should be used for standard PyTorch code?

Reply:
```python
pl_model = Trainer.compile(model, loss, optim, onnx=True)
```
Users will get a pl model, and users will use it as a pytorch model?

Emm, it's possible and (maybe) valid. I will check some details later.

Reply:
> ```python
> pl_model = Trainer.compile(model, loss, optim, onnx=True)
> ```
> 
> Users will get a pl model, and users will use it as a pytorch model?
> 
> Emm, it's possible and (maybe) valid. I will check some details later.

The goal is to transparently accelerate both PL and PyTorch code

Reply:
Surprisingly, it works well... (this example is from https://github.com/pytorch/examples/blob/master/mnist/main.py)
Only 4 lines to change and we can use the training speed up in bigdl-nano-init and inference speed up in onnxruntime
https://github.com/TheaperDeng/git_learning/compare/mnist-example-nano

Reply:
> Surprisingly, it works well... (this example is from [pytorch/examples@`master`/mnist/main.py](https://github.com/pytorch/examples/blob/master/mnist/main.py?rgh-link-date=2021-12-14T08%3A58%3A34Z)) Only 4 lines to change and we can use the training speed up in bigdl-nano-init and inference speed up in onnxruntime [TheaperDeng/git_learning@`mnist-example-nano` (compare)](https://github.com/TheaperDeng/git_learning/compare/mnist-example-nano)

What is `model.eval_onnx(tuple(next(iter(test_loader))[:-1]))`?

I think the below code might be problematic:

```python
model = Trainer.compile(model, nn.NLLLoss(), optimizer, onnx=True)

model.eval_onnx()
...

#model.exit_onnx() is not called
model.train()
...
```

And if the user just wants to use ONNX, it does not make sense to ask the user to specify loss and optimizer in `compile` (the user may not know what loss or optimizer is at all if he or she does not train the model).

Reply:
1. If users never call `Trainer.fit(model, data)` they need to put an example input in `eval_onnx`. I can accept a dataloader to be the input later, that is `model.eval_onnx(test_loader)`
2. Yes, I will check and find a way to make it work later...
3. We may allow some thing like this
```python
# model is a nn.module
model = Trainer.compile(model, onnx=True)
# use model like a nn.module
```